sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc

```diff
@@ -0,0 +1,458 @@
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Command group submission
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+
+}
+----
+
+[source,c++]
+----
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+----
+_Constraints_: Available only if `ExecutorType` is `sycl::queue`.
+
+_Effects_: Submit a command group function object to be scheduled for execution
+on `ex`. The submitted command will not execute until each of the events in
+`depEvents` is complete. An optional event representing this command can be
+returned via `outEvent`.
+
+
+=== Single tasks
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc);
+
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+
+}
+----
+
+[source,c++]
+----
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc);
+----
+_Constraints_: Available only if `ExecutorType` is `sycl::handler`.
+
+_Effects_: Enqueues a single task to the executor `ex`.
+
+
+[source,c++]
+----
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+----
+_Constraints_: Available only if `ExecutorType` is not `sycl::handler`.
+
+_Effects_: Enqueues a single task to the executor `ex`. The command will not
+execute until each of the events in `depEvents` is complete. An optional
+event representing this command can be returned via `outEvent`.
+
+
+=== Basic kernels
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, typename ExecutorType, int Dimensions, typename... Rest>
+void parallel_for(ExecutorType ex, range<Dimensions> r, Rest&&... rest);
+
+template <typename KernelName, typename ExecutorType, int Dimensions, typename... Rest>
+void parallel_for(ExecutorType ex, range<Dimensions> r, Rest&&... rest, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
```

Comment:
Can we have a parameter pack in the middle of the argument list like this?  According to this example, it seems like argument deduction doesn't work:

https://godbolt.org/z/dov75oa39

Reply:
Oh, that's annoying.  I tested the definition was legal, but clearly didn't test enough uses.

I guess we have to do what the SYCL 2020 specification does here, and instead define:
```c++
// Parameter pack acts as-if: Reductions&&... reductions, const KernelType &kernelFunc, const std::vector<event> = {}, event* outEvent = nullptr
template <typename KernelName, typename ExecutorType, int Dimensions, typename... Rest>
void parallel_for(ExecutorType ex, range<Dimensions> r, Rest&&... rest);
```

@rolandschulz, @steffenlarsen - Am I right in thinking we can do this?  We can use SFINAE for this?

Reply:
Using a parameter pack for both the "in" and "out" events make me worried about error messages.  For example, what sort of error message will we give if the user passes an `event` instead of an `event*` for the last parameter?

Reply:
I think the default error message wouldn't be _great,_ so we'd need to experiment with ways to make this better.

To get an idea for what might happen here, I tried to compile the below (where I'm passing an int instead of the expected kernel lambda):

```c++
q.parallel_for(1024, sycl::reduction((float*) nullptr, sycl::plus<>()), int());
```

Which produces:
```bash
reduction.hpp:2764:7: error: called object type '__tuple_element_t<1UL, tuple<reduction_impl<float, plus<void>, 0, 1, false, float *>, int>>' (aka 'int') is not a function or function pointer
      KernelFunc(GetDelinearized(I), Reducers...);
```

...which at least gives some clue that the int isn't a kernel function.

So, I expect we would get an error in the headers saying that it was an `event` instead of an `event*`.  But it wouldn't point directly to the user's call.

Reply:
Good error messages are important, so I think this should be factored into the API choice.  Maybe we should consider either "1b" (retrieve the "out" event via a property) or "1c" (retrieve the "out" event as a return value from a free function named like "_with_event").

Also, I just noticed that none of these APIs accept a property list at all.  This means we are losing functionality vs. [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc).

Reply:
I don't disagree that good error messages are important, but I don't think putting more things into the parameter pack makes things any worse.  We already need to find a way to generate better error messages for `parallel_for`, because we already have a parameter pack.  I think we should be able to do something here with `static_assert` -- I'll play around.

If we did go to "1b" or "1c", are you proposing that we'd make an equivalent change for all the other functions?  Or just `parallel_for`?

The property list being missing was an oversight, I'll add that back in.

Reply:
f72163d502a526b62917040928a7bee119d30a69 introduces the concept of a launch configuration to bundle some of these arguments together, as we discussed offline.